[SPARK-23936][SQL] Implement map_concat

bersprockets opened the pull request:
## What changes were proposed in this pull request?

Implement map_concat high order function.

This implementation does not pick a winner when the specified maps have overlapping keys. Therefore, this implementation preserves existing duplicate keys in the maps and potentially introduces new duplicates (After discussion with @ueshin, we settled on option 1 from [here](https://issues.apache.org/jira/browse/SPARK-23936?focusedCommentId=16464245&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16464245)).

## How was this patch tested?

New tests
Manual tests
Run all sbt SQL tests
Run all pyspark sql tests

bersprockets committed:
Initial commit

bersprockets committed:
Remove unused variable in test

bersprockets committed:
Cleanup

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #89369 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89369/testReport)** for PR 21073 at commit [`d04893b`](https://github.com/apache/spark/commit/d04893bccbd2772eb937125895519228588e74b4).

SparkQA commented:
**[Test build #89369 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89369/testReport)** for PR 21073 at commit [`d04893b`](https://github.com/apache/spark/commit/d04893bccbd2772eb937125895519228588e74b4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89369/
Test FAILed.

SparkQA commented:
**[Test build #89378 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89378/testReport)** for PR 21073 at commit [`97cffbe`](https://github.com/apache/spark/commit/97cffbea429d82accfcdece5e0fdcde903678cc8).

SparkQA commented:
**[Test build #89378 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89378/testReport)** for PR 21073 at commit [`97cffbe`](https://github.com/apache/spark/commit/97cffbea429d82accfcdece5e0fdcde903678cc8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89378/
Test PASSed.

henryr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -115,6 +116,62 @@ case class MapValues(child: Expression)
+      .getOrElse(MapType(keyType = StringType, valueType = StringType))
+  }
+
+  override def nullable: Boolean = false
```
What's the result of `map_concat(NULL, NULL)`?

bersprockets replied:
@henryr empty map:

<pre>
scala> df.select(map_concat('map1, 'map2).as('newMap)).show
df.select(map_concat('map1, 'map2).as('newMap)).show
+------+
|newMap|
+------+
|    []|
|    []|
+------+
</pre>
Presto docs (from which the proposed spec comes) are quiet on the matter. Even after looking at the Presto code, I am still hard-pressed to say.

I did divine from the Presto code that there should be at least two inputs (and I don't currently verify that).

henryr replied:
Hm, seems a bit unusual to me to have, in effect, `NULL ++ NULL => Map()`. I checked with Presto and it looks like it returns `NULL`:

    presto> select map_concat(NULL, NULL)
         -> ;
     _col0
    -------
     NULL
    (1 row)

bersprockets replied:
@henryr Since Presto is the reference, map_concat should return NULL in this case. I will update.

bersprockets replied:
@henryr Another quick test of Presto also shows that if *any* input is NULL, the result is NULL:
<pre>
presto:default> SELECT map_concat(NULL, map(ARRAY[1,3], ARRAY[2,4]));
 _col0 
-------
 NULL  
(1 row)
</pre>
Looks like I need to check if any input is NULL.

bersprockets committed:
Checkpoint non-working codegen

bersprockets committed:
Checkpoint somewhat working codegen

bersprockets committed:
Checkpoint better working codegen

bersprockets committed:
Require at least two input maps

bersprockets committed:
Small cleanup

bersprockets committed:
Remove redundant null check

SparkQA commented:
**[Test build #89473 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89473/testReport)** for PR 21073 at commit [`44137cc`](https://github.com/apache/spark/commit/44137cc9a9949b4218d973dc46d905d3ce301bcd).

SparkQA commented:
**[Test build #89473 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89473/testReport)** for PR 21073 at commit [`44137cc`](https://github.com/apache/spark/commit/44137cc9a9949b4218d973dc46d905d3ce301bcd).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89473/
Test PASSed.

SparkQA commented:
**[Test build #89523 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89523/testReport)** for PR 21073 at commit [`d3d6ad6`](https://github.com/apache/spark/commit/d3d6ad69bb00d58650120e58f4233f2ff6b40bdf).

SparkQA commented:
**[Test build #89523 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89523/testReport)** for PR 21073 at commit [`d3d6ad6`](https://github.com/apache/spark/commit/d3d6ad69bb00d58650120e58f4233f2ff6b40bdf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89523/
Test PASSed.

bersprockets committed:
Any null input means null result (ala Presto)

bersprockets committed:
Remove redundant null check

SparkQA commented:
**[Test build #89590 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89590/testReport)** for PR 21073 at commit [`a904c17`](https://github.com/apache/spark/commit/a904c179bbeef11492e1b0b3fe439400c3e4ddd0).

bersprockets commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +118,154 @@ case class MapValues(child: Expression)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val mapCodes = children.map(c => c.genCode(ctx))
```
Since this logic is big enough (and similar enough to the logic in eval), I wonder if the merge logic should be moved to a utility class and called from both eval as well as the generated code.

The FromUTCTimestamp expression does something sort of like that, where the eval method as well as the generated code both call utility functions in the DateTimeUtils scala object. Also, the Concat expression's eval method and generated code both call utility functions on UTF8String (although in this case, UTF8String is a Java class).

henryr replied:
FWIW, I don't really feel strongly either way here. The codegen method isn't so large as to be hard to understand yet.

gatorsmile commented:
cc @ueshin

SparkQA commented:
**[Test build #89590 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89590/testReport)** for PR 21073 at commit [`a904c17`](https://github.com/apache/spark/commit/a904c179bbeef11492e1b0b3fe439400c3e4ddd0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89590/
Test PASSed.

henryr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +118,154 @@ case class MapValues(child: Expression)
+    val vaName = ctx.freshName("va")
+    val keyName = ctx.freshName("key")
+    val valueName = ctx.freshName("value")
+    val isNullCheckName = ctx.freshName("isNull")
```
unused?

henryr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +118,154 @@ case class MapValues(child: Expression)
+        |for (int $index1Name = 0; $index1Name < $entrySetName.length; $index1Name++) {
+        |  $entryClass<Object, Object> entry =
+        |     ($entryClass<Object, Object>) $entrySetName[$index1Name];
+        |  $mergedKeyArrayName[$index1Name] = (Object) entry.getKey();
```
are the casts to `Object` necessary?

henryr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +118,154 @@ case class MapValues(child: Expression)
+      s"""
+        |$init
+        |$assignments
+        | if (!${ev.isNull}) {
```
is there one extra space before `if`?

henryr commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:
```diff
@@ -56,6 +58,26 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
+    checkEvaluation(MapConcat(Seq(m0, m1, m2)),
+      mutable.LinkedHashMap("a" -> "4", "b" -> "2", "c" -> "3", "d" -> "4", "e" -> "5"))
+    // null map
+    checkEvaluation(MapConcat(Seq(m0, mNull)),
```
good idea to check `Seq(mNull, m0)` as well in case there's any asymmetry in the way the first argument is handled.

bersprockets replied:
Done!

henryr commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -376,6 +376,35 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
+    assert(intercept[AnalysisException] {
+      df2.selectExpr("map_concat(map1, map2)").collect()
+    }.getMessage().contains("input maps of function map_concat should all be the same type"))
+    assert(intercept[AnalysisException] {
```
can you put a blank line between tests? makes it a bit easier to see the separation.

henryr commented on python/pyspark/sql/functions.py:
```diff
@@ -2186,6 +2186,29 @@ def map_values(col):
+    """
+    sc = SparkContext._active_spark_context
+    if len(cols) == 1 and isinstance(cols[0], (list, set)):
+        cols = cols[0]
```
what's this for?

bersprockets replied:
>what's this for?

Excellent question. I don't know, except that it seems sometimes the first column is a list of columns. I used other functions as a template.

henryr reviewed (commented):
This looks pretty good to me, would be good to have one of the people most familiar with codegen take a look.

bersprockets committed:
Review feedback

SparkQA commented:
**[Test build #89759 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89759/testReport)** for PR 21073 at commit [`13baf96`](https://github.com/apache/spark/commit/13baf96aa0a6087f288d45a7d57881744e187826).

SparkQA commented:
**[Test build #89759 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89759/testReport)** for PR 21073 at commit [`13baf96`](https://github.com/apache/spark/commit/13baf96aa0a6087f288d45a7d57881744e187826).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89759/
Test FAILed.

bersprockets commented:
retest this please

SparkQA commented:
**[Test build #89785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89785/testReport)** for PR 21073 at commit [`13baf96`](https://github.com/apache/spark/commit/13baf96aa0a6087f288d45a7d57881744e187826).

henryr commented:
@gatorsmile this looks ready for your review (asking because you filed the JIRA) if you time, thanks!

mn-mikke commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +118,153 @@ case class MapValues(child: Expression)
+       [[1 -> "a"], [2 -> "c"], [3 -> "d"]
+  """)
+case class MapConcat(children: Seq[Expression]) extends Expression
+  with CodegenFallback {
```
Do you need to inherit from `CodegenFallback`  if you've overriden `doGenCode`?

mn-mikke commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +118,153 @@ case class MapValues(child: Expression)
+        |  $arrayDataClass $vaName = $mapDataName.valueArray();
+        |  for (int $index2Name = 0; $index2Name < $kaName.numElements(); $index2Name++) {
+        |    Object $keyName = ${CodeGenerator.getValue(kaName, keyType, index2Name)};
+        |    Object $valueName = ${CodeGenerator.getValue(vaName, valueType, index2Name)};
```
I think you should add handling of `null`s when values are of a primitive type.

mn-mikke commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +118,153 @@ case class MapValues(child: Expression)
+    Examples:
+      > SELECT _FUNC_(map(1, 'a', 2, 'b'), map(2, 'c', 3, 'd'));
+       [[1 -> "a"], [2 -> "c"], [3 -> "d"]
+  """)
```
since = "2.4.0"

mn-mikke commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:
```diff
@@ -56,6 +58,28 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
     checkEvaluation(MapValues(m2), null)
   }
 
+  test("Map Concat") {
```
Please add more test cases with `null` values.

bersprockets replied:
Done.

bersprockets committed:
Check for null value in generated code

bersprockets committed:
Add since to expression description

bersprockets committed:
Allow valueContainsNull to vary; Make checkInputDataTypes more in line with other expressions

bersprockets committed:
Add a few more tests

bersprockets committed:
One more test

SparkQA commented:
**[Test build #89944 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89944/testReport)** for PR 21073 at commit [`2e49b1e`](https://github.com/apache/spark/commit/2e49b1e01ba10d7baba9196d64af8db1cd7b2dd1).

SparkQA commented:
**[Test build #89944 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89944/testReport)** for PR 21073 at commit [`2e49b1e`](https://github.com/apache/spark/commit/2e49b1e01ba10d7baba9196d64af8db1cd7b2dd1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89944/
Test PASSed.

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -16,12 +16,14 @@
 import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
 import org.apache.spark.sql.catalyst.expressions.codegen._
 import org.apache.spark.sql.catalyst.util.{ArrayData, GenericArrayData, MapData, TypeUtils}
+import org.apache.spark.sql.catalyst.util.ArrayBasedMapData
```
How about merging these two lines into one line `org.apache.spark.sql.catalyst.util._`?

bersprockets committed:
Fix import statement; Add two small tests

bersprockets commented:
@mn-mikke @kiszk Thanks for the review. I addressed the comments. Please take a look when you have a chance.

bersprockets commented:
retest this please

bersprockets commented:
retest this please

SparkQA commented:
**[Test build #89993 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89993/testReport)** for PR 21073 at commit [`d9dccd3`](https://github.com/apache/spark/commit/d9dccd3c96a5d1ffdfce0225828af21bd1ee8303).

SparkQA commented:
**[Test build #89993 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89993/testReport)** for PR 21073 at commit [`d9dccd3`](https://github.com/apache/spark/commit/d9dccd3c96a5d1ffdfce0225828af21bd1ee8303).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89993/
Test FAILed.

bersprockets commented:
A test failed with "./bin/spark-submit ... No such file or directory"

Seems like there's lots of spurious test failures right now. I will hold off on re-running for a little while.

bersprockets commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +117,161 @@ case class MapValues(child: Expression)
+      val map = raw.asInstanceOf[MapData]
+      map.foreach(dataType.keyType, dataType.valueType, (k, v) =>
+        union.put(k, v)
+      )
```
I found an issue. I was preparing to add some more tests when I noticed that using maps as keys doesn't work well in interpreted mode (seems to work fine in codegen mode, so far).

So, something like this doesn't work in interpreted mode:
<pre>
scala> dfmapmap.show(truncate=false)
+--------------------------------------------------+---------------------------------------------+
|mapmap1                                           |mapmap2                                      |
+--------------------------------------------------+---------------------------------------------+
|[[1 -> 2, 3 -> 4] -> 101, [5 -> 6, 7 -> 8] -> 102]|[[11 -> 12] -> 103, [1 -> 2, 3 -> 4] -> 1001]|
+--------------------------------------------------+---------------------------------------------+
scala> dfmapmap.select(map_concat('mapmap1, 'mapmap2).as('mapmap3)).show(truncate=false)
+-----------------------------------------------------------------------------------------------+
|mapmap3                                                                                        |
+-----------------------------------------------------------------------------------------------+
|[[1 -> 2, 3 -> 4] -> 101, [5 -> 6, 7 -> 8] -> 102, [11 -> 12] -> 103, [1 -> 2, 3 -> 4] -> 1001]|
+-----------------------------------------------------------------------------------------------+
</pre>
As you can see, the key `[1 -> 2, 3 -> 4]` shows up twice in the new map.

This is because:
<pre>
  val a1 = new ArrayBasedMapData(new GenericArrayData(Array(1, 3)), new GenericArrayData(Array(2, 4)))
  val a2 = new ArrayBasedMapData(new GenericArrayData(Array(1, 3)), new GenericArrayData(Array(2, 4)))
  a1 == a2 // will be false
  a1.hashCode() == a2.hashCode() // will be false
</pre>
Different instances of ArrayBasedMapData with the exact same data are not considered the same key.

So far, seems to work fine for instances of UnsafeMapData (that is, the UnsafeMapData version of the above maps are considered the same key).

I might need to add equals and/or hashCode methods to ArrayBasedMapData. Either that, or wrap any ArrayBasedMapData with some other object (only during its time as a key in the union Map) that does implement equals and/or hashCode properly.

ueshin replied:
@bersprockets Hi, thanks for the investigation. We don't need to care about key duplication like `CreateMap` for now.

SparkQA commented:
**[Test build #90020 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90020/testReport)** for PR 21073 at commit [`d9dccd3`](https://github.com/apache/spark/commit/d9dccd3c96a5d1ffdfce0225828af21bd1ee8303).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90020/
Test FAILed.

bersprockets committed:
As per SPARK-9415, cannot compare Maps, therefore cannot support them as keys

SparkQA commented:
**[Test build #90091 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90091/testReport)** for PR 21073 at commit [`77ae014`](https://github.com/apache/spark/commit/77ae014bea3d9fbc20fbd11d5508c3606e26343d).

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +117,169 @@ case class MapValues(child: Expression)
+      val e2 = e.asInstanceOf[java.util.Map.Entry[Any, Any]]
+      (e2.getKey, e2.getValue)
+    }.unzip
+    new ArrayBasedMapData(new GenericArrayData(keyArray), new GenericArrayData(valueArray))
```
nit: we can use `ArrayBasedMapData.apply()`.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +117,169 @@ case class MapValues(child: Expression)
+      """.stripMargin
+
+    val assignments = mapCodes.zipWithIndex.map { case (m, i) =>
+      val initCode = mapCodes(i).code
```
`m.code`?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +117,169 @@ case class MapValues(child: Expression)
+
+    val assignments = mapCodes.zipWithIndex.map { case (m, i) =>
+      val initCode = mapCodes(i).code
+      val valueVarName = mapCodes(i).value.code
```
`m.value`?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +117,169 @@ case class MapValues(child: Expression)
+         |  ${ev.isNull} = true;
+         |}
+       """.stripMargin
+    }.mkString("\n")
```
Use `ctx.splitExpressionsWithCurrentInputs()` or something to avoid exceeding JVM limit.

SparkQA commented:
**[Test build #90091 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90091/testReport)** for PR 21073 at commit [`77ae014`](https://github.com/apache/spark/commit/77ae014bea3d9fbc20fbd11d5508c3606e26343d).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90091/
Test FAILed.

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -116,6 +117,169 @@ case class MapValues(child: Expression)
+        union.put(k, v)
+      )
+    }
+    val (keyArray, valueArray) = union.entrySet().toArray().map { e =>
```
What happens if union.entrySet().toArray() has more than 0x7FFF_FFFF elements?

bersprockets replied:
I would imagine bad things would happen before you got this far (even Map's size method returns an Int).

bersprockets committed:
Updates for some review comments

bersprockets committed:
Initial commit of use of splitExpressionsWithCurrentInputs

bersprockets committed:
Add test

bersprockets committed:
Fix indentation

SparkQA commented:
**[Test build #90320 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90320/testReport)** for PR 21073 at commit [`5b5e07e`](https://github.com/apache/spark/commit/5b5e07e0218e6a8480e167d9a03d50d0da0331fa).

SparkQA commented:
**[Test build #90320 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90320/testReport)** for PR 21073 at commit [`5b5e07e`](https://github.com/apache/spark/commit/5b5e07e0218e6a8480e167d9a03d50d0da0331fa).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90320/
Test PASSed.

SparkQA commented:
**[Test build #90336 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90336/testReport)** for PR 21073 at commit [`4c996d8`](https://github.com/apache/spark/commit/4c996d8328fe903067c4e8b5ed3bea70f73d7bb5).

bersprockets commented:
@ueshin Hopefully I have addressed all of your review comments.

Also, I have a question about what it means to dedup across maps when Spark allows duplicates in maps [here.](https://issues.apache.org/jira/browse/SPARK-23936?focusedCommentId=16464245&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16464245)

SparkQA commented:
**[Test build #90774 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90774/testReport)** for PR 21073 at commit [`eab0df2`](https://github.com/apache/spark/commit/eab0df25681778c3f351d6e0e2fcbfcf4cf2f37d).

SparkQA commented:
**[Test build #90774 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90774/testReport)** for PR 21073 at commit [`eab0df2`](https://github.com/apache/spark/commit/eab0df25681778c3f351d6e0e2fcbfcf4cf2f37d).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90774/
Test FAILed.

bersprockets commented:
retest this please

SparkQA commented:
**[Test build #90792 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90792/testReport)** for PR 21073 at commit [`eab0df2`](https://github.com/apache/spark/commit/eab0df25681778c3f351d6e0e2fcbfcf4cf2f37d).

SparkQA commented:
**[Test build #90792 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90792/testReport)** for PR 21073 at commit [`eab0df2`](https://github.com/apache/spark/commit/eab0df25681778c3f351d6e0e2fcbfcf4cf2f37d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90792/
Test PASSed.

bersprockets committed:
Fix after rebase

SparkQA commented:
**[Test build #91125 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91125/testReport)** for PR 21073 at commit [`a45e36f`](https://github.com/apache/spark/commit/a45e36fa775ac233eec10469058776b10746c6ec).

SparkQA commented:
**[Test build #91125 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91125/testReport)** for PR 21073 at commit [`a45e36f`](https://github.com/apache/spark/commit/a45e36fa775ac233eec10469058776b10746c6ec).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91125/
Test PASSed.

SparkQA commented:
**[Test build #91348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91348/testReport)** for PR 21073 at commit [`81ea4d9`](https://github.com/apache/spark/commit/81ea4d97e30672cabdd445eb079daa3842351520).

SparkQA commented:
**[Test build #91348 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91348/testReport)** for PR 21073 at commit [`81ea4d9`](https://github.com/apache/spark/commit/81ea4d97e30672cabdd445eb079daa3842351520).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91348/
Test PASSed.

bersprockets committed:
Review feedback: use pre-existing empty collections

SparkQA commented:
**[Test build #91364 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91364/testReport)** for PR 21073 at commit [`ee9ad54`](https://github.com/apache/spark/commit/ee9ad54413c2e74b011b60caf1b27ab8847741d2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91364/
Test PASSed.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -308,6 +308,170 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+ * Returns the union of all the given maps.
+ */
+@ExpressionDescription(
+usage = "_FUNC_(map, ...) - Returns the union of all the given maps",
```
nit: indent

bersprockets replied:
@ueshin 

>We don't need to care about key duplication like CreateMap for now.

Just verifying: This means I should simply concatenate the maps, possibly creating additional duplicates.

ueshin replied:
Yes, currently we don't check the key duplication in other functions as well. We should follow the behavior for now. We might need to discuss the issue in the future, so let's revisit the issue here at that time.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -308,6 +308,170 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+examples = """
+    Examples:
+      > SELECT _FUNC_(map(1, 'a', 2, 'b'), map(2, 'c', 3, 'd'));
+       [[1 -> "a"], [2 -> "c"], [3 -> "d"]
```
The brackets doesn't match?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -308,6 +308,170 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+  }
+
+  override def dataType: MapType = {
+    MapType(
```
How about doing `children.map(_.dataType.asInstanceOf[MapType])` first?

AmplabJenkins commented:
Can one of the admins verify this patch?

bersprockets committed:
Allow duplicate keys

SparkQA commented:
**[Test build #92147 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92147/testReport)** for PR 21073 at commit [`58b2b9d`](https://github.com/apache/spark/commit/58b2b9d5958bf0d9ff677423fd304bd3356a1b3b).

SparkQA commented:
**[Test build #92147 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92147/testReport)** for PR 21073 at commit [`58b2b9d`](https://github.com/apache/spark/commit/58b2b9d5958bf0d9ff677423fd304bd3356a1b3b).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92147/
Test FAILed.

SparkQA commented:
**[Test build #92152 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92152/testReport)** for PR 21073 at commit [`3e92f4b`](https://github.com/apache/spark/commit/3e92f4b2e03771a18a53fd35eee72ad811a5caf3).

SparkQA commented:
**[Test build #92152 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92152/testReport)** for PR 21073 at commit [`3e92f4b`](https://github.com/apache/spark/commit/3e92f4b2e03771a18a53fd35eee72ad811a5caf3).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92152/
Test FAILed.

bersprockets committed:
Remove extra line added during rebase

SparkQA commented:
**[Test build #92153 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92153/testReport)** for PR 21073 at commit [`1b61212`](https://github.com/apache/spark/commit/1b612129de771ce30a8bfe1a9459899abfa7d46a).

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -474,6 +473,221 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+    val assignments = mapCodes.zipWithIndex.map { case (m, i) =>
+      s"""
+         |${m.code}
+         |$argsName[$i] = ${m.value.code};
```
Can you check null while assigning children? We don't need to evaluate the rest of children if we find there is a null child. I remember `ArraysZip` is doing the similar thing.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -474,6 +473,221 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+       |    int $counter = 0;
+       |    for (int y = 0; y < ${children.length}; y++) {
+       |      for (int z = 0; z < $argsName[y].numElements(); z++) {
+       |        if ($argsName[y].isNullAt(z)) {
```
We don't need to check null for keys, and for values if `!valueContainsNull`.

ueshin commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:
```diff
@@ -80,6 +80,138 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
+    assert(MapConcat(Seq(m5, m6)).dataType.valueContainsNull)
+    assert(MapConcat(Seq(m6, m5)).dataType.valueContainsNull)
+    assert(!MapConcat(Seq(m1, m2)).nullable)
+    assert(MapConcat(Seq(m1, mNull)).nullable)
```
These checks should be done in `DataFrameFunctionsSuite`?

ueshin commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:
```diff
@@ -80,6 +80,138 @@ class CollectionExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper
+
+    // null map
+    checkEvaluation(MapConcat(Seq(m0, mNull)),
+      null)
```
This should be 1 line? Also the following lines.

ueshin commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:
```diff
@@ -633,6 +633,51 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
+        Row(Map(1 -> 400, 2 -> 200, 3 -> 300)),
+        Row(null)
+      )
+    )
```
Can you add tests using `select(map_concat($"map1", $"map2"))` as well?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -474,6 +473,221 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+    val codes = ctx.splitExpressionsWithCurrentInputs(
+      expressions = assignments,
+      funcName = "mapConcat",
+      extraArguments = (s"${mapDataClass}[]", argsName) :: Nil)
```
nit: `s"$mapDataClass[]"`.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -474,6 +473,221 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+      TypeCheckResult.TypeCheckFailure(
+        s"The given input of function $prettyName should all be of type map, " +
+          "but they are " + children.map(_.dataType.simpleString).mkString("[", ", ", "]"))
+    } else if (children.map(_.dataType.asInstanceOf[MapType].keyType).distinct.length > 1) {
```
We can use `TypeUtils.checkForSameTypeInputExpr()`?

ueshin reviewed (commented):
I left some comments.
I'm wondering whether we need type coercion like `concat` for array type is doing.

bersprockets commented:
Hi @ueshin. Just a question while I work on the changes for your review comments.
>I'm wondering whether we need type coercion like concat for array type is doing.

Which type coercion in Concat are you referring to? Are you referring to the check for primitive vs object type in code generation?

bersprockets committed:
Review comments

SparkQA commented:
**[Test build #92269 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92269/testReport)** for PR 21073 at commit [`ce5541e`](https://github.com/apache/spark/commit/ce5541edaa1d3eb8a2632c09c49bd89c79ce7fe6).

SparkQA commented:
**[Test build #92269 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92269/testReport)** for PR 21073 at commit [`ce5541e`](https://github.com/apache/spark/commit/ce5541edaa1d3eb8a2632c09c49bd89c79ce7fe6).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92269/
Test PASSed.

SparkQA commented:
**[Test build #92273 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92273/testReport)** for PR 21073 at commit [`4893df9`](https://github.com/apache/spark/commit/4893df9c88223ee1bb9d78d9c7882fbd4a56319e).

SparkQA commented:
**[Test build #92273 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92273/testReport)** for PR 21073 at commit [`4893df9`](https://github.com/apache/spark/commit/4893df9c88223ee1bb9d78d9c7882fbd4a56319e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92273/
Test PASSed.

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -475,6 +474,231 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+      s"""
+        |$mapDataClass[] $argsName = new $mapDataClass[${mapCodes.size}];
+        |$arrayDataClass[] $keyArgsName = new $arrayDataClass[${mapCodes.size}];
+        |$arrayDataClass[] $valArgsName = new $arrayDataClass[${mapCodes.size}];
```
We should create these two arrays later?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -475,6 +474,231 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+    val codes = ctx.splitExpressionsWithCurrentInputs(
+      expressions = assignments,
+      funcName = "getMapConcatInputs",
+      extraArguments = (s"$mapDataClass[]", argsName) :: ("boolean", ev.isNull.code) :: Nil,
```
`ev.isNull` instead of `ev.isNull.code`?

bersprockets replied:
<code>ev.isNull</code> does not compile for me.
<pre>
[error]  found   : List[(String, Object)]
[error]  required: Seq[(String, String)]
[error]       extraArguments = (s"$mapDataClass[]", argsName) :: ("boolean", ev.isNull) :: Nil,
</pre>
<code>ev.isNull.code</code> and also <code>ev.isNull.toString</code> does compile.

ueshin replied:
I see. Can you add another variable to store if `null` exists, and set it to `${ev.isNull}` later, just in case?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -475,6 +474,231 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+
+    val assignments = mapCodes.zipWithIndex.map { case (m, i) =>
+      s"""
+         |${m.code}
```
We should check `${ev.isNull}` first and we can skip evaluation of the rest of the children if it's `true`?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -475,6 +474,231 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+  }
+
+  private def genCodeForPrimitiveArrays(ctx: CodegenContext, elementType: DataType,
+                                        checkForNull: Boolean): String = {
```
nit: style.

```scala
private def genCodeForPrimitiveArrays(
    ctx: CodegenContext,
    elementType: DataType,
    checkForNull: Boolean): String = {
```

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -475,6 +474,231 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+         |$arrayData.set$primitiveValueTypeName(
+         |  $counter,
+         |  ${CodeGenerator.getValue(s"$argsName[y]", elementType, "z")}
+         |);""".stripMargin.stripPrefix("\n")
```
Do we need `.stripPrefix("\n")` here?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -475,6 +474,231 @@ case class MapEntries(child: Expression) extends UnaryExpression with ExpectsInp
+         |  $arrayData.setNullAt($counter);
+         |} else {
+         |  $setterCode1
+         |}""".stripMargin.stripPrefix("\n")
```
ditto.

ueshin reviewed (commented):
@bersprockets I meant, `Concat` for array type is doing type coercion at [TypeCoercion.scala#L531-L537](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L531-L537), so I was wondering whether we need the same thing for `MapConcat` or not.

bersprockets commented:
@ueshin 
>so I was wondering whether we need the same thing for MapConcat or not.

Got it. I will research that, plus I will look at the entire pull request for Concat to see if there is anything else relevant to MapConcat.

bersprockets committed:
Review comments

SparkQA commented:
**[Test build #92356 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92356/testReport)** for PR 21073 at commit [`a899065`](https://github.com/apache/spark/commit/a89906524abad0d7b70307e7afef67b81e379e0c).

bersprockets commented:
Still working on type coercion.

SparkQA commented:
**[Test build #92356 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92356/testReport)** for PR 21073 at commit [`a899065`](https://github.com/apache/spark/commit/a89906524abad0d7b70307e7afef67b81e379e0c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92356/
Test PASSed.

bersprockets committed:
Initial implementation of type coercion for map_concat

SparkQA commented:
**[Test build #92502 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92502/testReport)** for PR 21073 at commit [`484ac17`](https://github.com/apache/spark/commit/484ac173d253efcf83a86edc02f61cc3cba57d02).

SparkQA commented:
**[Test build #92502 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92502/testReport)** for PR 21073 at commit [`484ac17`](https://github.com/apache/spark/commit/484ac173d253efcf83a86edc02f61cc3cba57d02).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92502/
Test PASSed.

SparkQA commented:
**[Test build #92505 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92505/testReport)** for PR 21073 at commit [`4ee7b46`](https://github.com/apache/spark/commit/4ee7b46ef7d337c37b6198104ea37a93a5584e72).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92505/
Test PASSed.

kiszk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -17,6 +17,7 @@
 package org.apache.spark.sql.catalyst.expressions
 
 import java.util.{Comparator, TimeZone}
+import java.util
```
nit: Do we need this import?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -551,6 +551,36 @@ object TypeCoercion {
 
+      case m @ MapConcat(children) if children.forall(c => MapType.acceptsType(c.dataType)) &&
+        !haveSameType(children) =>
+        val keyTypes = children.map(_.dataType.asInstanceOf[MapType].keyType)
```
Can't we do the same thing as the `Concat` does?

```scala
val types = children.map(_.dataType)
findWiderCommonType(types) match {
  case Some(finalDataType) => MapConcat(children.map(Cast(_, finalDataType)))
  case None => m
}
```

bersprockets replied:
I don't necessarily have a _good_ reason, but here are two reasons I did that extra junk:

1) The Concat-like code didn't find a wider type amongst types map<tinyint,tinyint> and map<int,int>. So, it just fell to <code>case None => m</code>

2) If the call to map_concat has the same child types but multiple valueContainsNull values, the Concat-style code added a Cast to each child (this is because haveSameType considers expressions with different valueContainsNull values to have different types). It does no harm, as far as I can tell, but it seemed wrong.

About issue 1): I will debug. I might have done something wrong there. Plus, even if it's a real bug in findWiderCommonType, it affects my longer code, which may be looking for a wider common type amongst the keys or values, which could themselves be maps.

About issue 2): Maybe not an issue. Or I can create an alternate haveSameType() function for Maps.

ueshin commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:
```diff
@@ -20,6 +20,8 @@ package org.apache.spark.sql.catalyst.expressions
 import java.sql.{Date, Timestamp}
 import java.util.TimeZone
 
+import scala.collection.mutable
```
Unnecessary import?

SparkQA commented:
**[Test build #92660 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92660/testReport)** for PR 21073 at commit [`3c0da03`](https://github.com/apache/spark/commit/3c0da039a0f11e0bc4f18342c96cf1dc100d2060).

bersprockets committed:
Simplify type coercion for map_concat parameters

SparkQA commented:
**[Test build #92662 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92662/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).

SparkQA commented:
**[Test build #92662 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92662/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92662/
Test FAILed.

SparkQA commented:
**[Test build #92660 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92660/testReport)** for PR 21073 at commit [`3c0da03`](https://github.com/apache/spark/commit/3c0da039a0f11e0bc4f18342c96cf1dc100d2060).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92660/
Test PASSed.

ueshin commented:
LGTM pending Jenkins.

SparkQA commented:
**[Test build #92672 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92672/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).

bersprockets commented:
retest this please.

SparkQA commented:
**[Test build #92689 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92689/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).

SparkQA commented:
**[Test build #92689 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92689/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92689/
Test PASSed.

ueshin commented:
I'd retrigger the build, just in case.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #92728 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92728/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).

SparkQA commented:
**[Test build #92728 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92728/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92728/
Test FAILed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #92733 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92733/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).

SparkQA commented:
**[Test build #92733 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92733/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92733/
Test FAILed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #92740 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92740/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).

SparkQA commented:
**[Test build #92740 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92740/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92740/
Test FAILed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #92745 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/92745/testReport)** for PR 21073 at commit [`03328a4`](https://github.com/apache/spark/commit/03328a417ea04722c1497cf09583dff909afe979).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/92745/
Test PASSed.

ueshin commented:
Thanks! merging to master.

bersprockets commented:
@ueshin Thanks for all your help!
